examples/async-sqlite/src/lib.rs

```diff
@@ -69,155 +96,48 @@ impl Database {
-        let id = cx.argument::<JsNumber>(0)?.value(&mut cx);
-
-        // Get the `this` value as a `JsBox<Database>`
-        let db = cx.this::<JsBox<Database>>()?;
-        let (deferred, promise) = cx.promise();
-
-        db.send(deferred, move |conn, channel, deferred| {
-            let result: Result<String, _> = conn
-                .prepare("SELECT name FROM person WHERE id = ?")
-                .and_then(|mut stmt| stmt.query_row(rusqlite::params![id], |row| row.get(0)));
-
-            deferred.settle_with(channel, move |mut cx| -> JsResult<JsValue> {
-                // If the row was not found, return `undefined` as a success instead
-                // of throwing an exception
-                if matches!(result, Err(rusqlite::Error::QueryReturnedNoRows)) {
-                    return Ok(cx.undefined().upcast());
-                }
-
-                let name = result.or_else(|err| cx.throw_error(err.to_string()))?;
-
-                Ok(cx.string(name).upcast())
-            });
-        })
-        .into_rejection(&mut cx)?;
-
-        Ok(promise)
-    }
-}
-
-trait SendResultExt {
-    // Sending a query closure to execute may fail if the channel has been closed.
-    // This method converts the failure into a promise rejection.
-    fn into_rejection<'a, C: Context<'a>>(self, cx: &mut C) -> NeonResult<()>;
-}
-
-impl SendResultExt for Result<(), mpsc::SendError<DbMessage>> {
-    fn into_rejection<'a, C: Context<'a>>(self, cx: &mut C) -> NeonResult<()> {
-        self.or_else(|err| {
-            let msg = err.to_string();
-
-            match err.0 {
-                DbMessage::Callback(deferred, _) => {
-                    let err = cx.error(msg)?;
-                    deferred.reject(cx, err);
-                    Ok(())
-                }
-                DbMessage::Close => cx.throw_error("Expected DbMessage::Callback"),
+    fn by_id<'cx>(&self, cx: &mut Cx<'cx>, id: f64) -> JsResult<'cx, JsPromise> {
+        self.exec(cx, move |conn| -> Result<Option<String>, Error> {
+            match conn
+                .prepare("SELECT name FROM person WHERE id = ?")?
+                .query_row(rusqlite::params![id], |row| row.get(0))
+            {
+                Ok(name) => Ok(Some(name)),
+                Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
+                Err(err) => Err(Error::from(err)),
             }
         })
     }
-}
-
-#[neon::main]
-fn main(mut cx: ModuleContext) -> NeonResult<()> {
-    cx.export_function("databaseNew", Database::js_new)?;
-    cx.export_function("databaseClose", Database::js_close)?;
-    cx.export_function("databaseInsert", Database::js_insert)?;
-    cx.export_function("databaseGetById", Database::js_get_by_id)?;
 
-    Ok(())
+    // Idiomatic rust would take an owned `self` to prevent use after close
+    // However, it's not possible to prevent JavaScript from continuing to hold a closed database
```

Comment:
What happens if they call methods on a closed database? Should we wrap `tx` in an `Option`?

Reply:
OK I **think** I get how this works, but it's incredibly subtle. Check my understanding:
- When an operation fails or the user calls `.close()` for the first time, the operation produces `ControlFlow::Break()`
- This causes the database thread's main loop to terminate, and the thread terminates
- The thread owns the mpsc receiver, so it gets dropped
- Subsequent attempts to use the sender will fail because the mpsc semantics is that when one half is dropped, most operations on the other half fail from that point on

It would be helpful to explain at least some of this in the comments.